php/db.php

```diff
@@ -73,6 +76,103 @@ function GetDataForTable($tabName){
 	return $tabData;
 }
 
+function MigrateDatabase() {
+	global $dbConn, $ip, $userAgent, $dbVersion;
+	$migrationsDir = "SQL/versions/";
+
+
+	StartTimer("MigrateDatabase");
+
+	// Check our current database version
+	$sql = "SELECT config_value FROM config WHERE config_key = 'DATABASE_VERSION';";
+	$data = mysqli_query($dbConn, $sql);
+	
+	$config_result = mysqli_fetch_array($data);
+
+	if ($config_result == NULL) {
+		die("Unable to determine the database version. Please add the config value DATABASE_VERSION before running");
+	}
+
+	$currentDatabaseVersion = intval($config_result[0], 10);
+
+	// Check and see if we need to migrate.
+	if ($dbVersion < $currentDatabaseVersion) {
+		die('Database version is out of date with code version');
+	} elseif ($dbVersion == $currentDatabaseVersion) {
+		StopTimer("MigrateDatabase");
+		return;
+	}
+	
+	// Calculate what migrations exist
+	$migrations = array();
+	
+	$migrationDirFiles = scandir($migrationsDir);
+	foreach ($migrationDirFiles as $i => $migrationFile) {
+		// Get the version ID from the filename. Files need to be formatted <versionID>_<randomName>.sql
+		preg_match('/(\d+)_.*\.sql$/', $migrationFile, $matches);
+		if (isset($matches[1])) {
+			$migrationID = intval($matches[1], 10);
+			$migrations[$migrationID] = $migrationFile;
+		}
+	}
+
+	// Go through all of the needed migrations and run them
+	for($i = $currentDatabaseVersion+1; $i <= $dbVersion; $i++) {
+
+		// Make sure that migration ID actually exists. This is so we can have version 10, 11, 14, (notice the missing 13?)
+		if (isset($migrations[$i])) {
+			$migrationScript = $migrations[$i];
+
+			// Update our adminlog so we can see what happened before the migration
+			$escapedIP = mysqli_real_escape_string($dbConn, $ip);
+			$escapedUserAgent = mysqli_real_escape_string($dbConn, $userAgent);
+			$escapedLog = mysqli_real_escape_string($dbConn, "Running $migrationScript to bring DATABASE_VERSION up to $i");
+			$sql = "
+				INSERT INTO admin_log
+				(log_id, log_datetime, log_ip, log_user_agent, log_admin_username, log_subject_username, log_type, log_content)
+				VALUES
+				(
+					null,
+					Now(),
+					'$escapedIP',
+					'$escapedUserAgent',
+					'AUTOMATIC',
+					'',
+					'DB_MIGRATION',
+					'$escapedLog'
+				);";
+		
+			mysqli_query($dbConn, $sql) or die("Migration failed to log to admin log, please notify site admin");;
+			$sql = "";
+
+			// Run the script
+			$sql = file_get_contents("$migrationsDir/$migrationScript");
+			if(mysqli_multi_query($dbConn, $sql)) {
+				do {
+					mysqli_next_result($dbConn);
+				}
+				while(mysqli_more_results($dbConn));
```

Comment:
Please add a safety here (and in every while or do-while) to prevent infinite loops, which triggers a die("Infinite loop in MigrateDatabase"); or something if it iterates too many times: `$safety = 100;` -> `$safety--; if($safety <= 0){die(..);}`

Reply:
This is the main thing I disagree with. I think the safety isn't needed as far as I can tell this is the defacto way of doing multi-queries with mysqli. For example the example #1 here https://www.php.net/manual/en/mysqli.multi-query.php shows the use of the do while. I think adding safety's to code that is running in a closed ecosystem (so not relying on user input to determine the ending condition) is unnecessary and confusing to future readers of the code.